src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -861,6 +867,315 @@ bool canUseParallelReplicasOnInitiator(const ContextPtr & context)
+    for (size_t i = 0, s = pools_to_use.size(); i < s; ++i)
+    {
+        const auto & hostname = pools_to_use[i]->getHost();
+        const auto found = std::find_if(
+            begin(shard.local_addresses),
+            end(shard.local_addresses),
+            [&hostname](const Cluster::Address & local_addr) { return hostname == local_addr.host_name; });
+        if (found != shard.local_addresses.end())
+        {
+            local_replica_index = i;
+            break;
+        }
+    }
+    if (!local_replica_index)
+        throw Exception(
+            ErrorCodes::INCONSISTENT_CLUSTER_DEFINITION,
+            "Local replica is not found in '{}' cluster definition, see 'cluster_for_parallel_replicas' setting",
+            new_cluster->getName());
+
+    // resize the pool but keep local replicas in it (and update its index)
+    chassert(max_replicas_to_use <= pools_to_use.size());
+    if (local_replica_index >= max_replicas_to_use)
+    {
+        std::swap(pools_to_use[max_replicas_to_use - 1], pools_to_use[local_replica_index.value()]);
+        local_replica_index = max_replicas_to_use - 1;
+    }
+    pools_to_use.resize(max_replicas_to_use);
+
+    LOG_DEBUG(logger, "Local replica got replica number {}", local_replica_index.value());
+
+    String formatted_query;
+    {
+        InterpreterSelectQueryAnalyzer analyzer(query_ast.select, context, {});
+        const auto & query_tree = analyzer.getQueryTree();
+        auto select_ast = query_tree->toAST();
+
+        auto new_query_ast = query_ast.clone();
+        auto * insert_ast = new_query_ast->as<ASTInsertQuery>();
+        insert_ast->select = std::move(select_ast);
+
+        WriteBufferFromOwnString buf;
+        IAST::FormatSettings ast_format_settings(
+            /*one_line=*/true, /*hilite=*/false, /*identifier_quoting_rule=*/IdentifierQuotingRule::Always);
+        insert_ast->IAST::format(buf, ast_format_settings);
+        formatted_query = buf.str();
+    }
+
+    const bool skip_local_replica = local_pipeline.has_value();
+    QueryPipeline pipeline;
+    if (local_pipeline)
+    {
+        chassert(coordinator.has_value());
+        pipeline.addCompletedPipeline(std::move(*local_pipeline));
+    }
+    if (!coordinator)
+        coordinator = std::make_shared<ParallelReplicasReadingCoordinator>(max_replicas_to_use);
+
+    for (size_t i = 0; i < pools_to_use.size(); ++i)
+    {
+        if (skip_local_replica && i == *local_replica_index)
+            continue;
+
+        IConnections::ReplicaInfo replica_info{
+            /// we should use this number specifically because efficiency of data distribution by consistent hash depends on it.
+            .number_of_current_replica = i,
+        };
+        auto remote_query_executor = std::make_shared<RemoteQueryExecutor>(
+            pools_to_use[i],
+            formatted_query,
+            Block{},
+            new_context,
+            /*throttler=*/nullptr,
```

Comment:
Why not pass throttler here?

Reply:
Initially no reason. But we don't transfer table data over network in this case. Only, some coordination packets

Reply:
Make sense, and it is indeed does not worth to pass it here, let's also add a comment about this, so that it won't appeared when someone will do some massive refactoring